### PR TITLE
8307181: MemoryLayout.structLayout uses undocumented strict alignment constraints

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch adds documentation for the behavior of the `MemoryLayout::structLayout` factory.

This factory throws an `IllegalArgumentException` if one of the member layouts passed to it occurs at an offset within the struct that is not aligned.

As part of this patch, I've also added some api notes on how to use this factory when unaligned elements are present.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8307335](https://bugs.openjdk.org/browse/JDK-8307335) to be approved

### Issues
 * [JDK-8307181](https://bugs.openjdk.org/browse/JDK-8307181): MemoryLayout.structLayout uses undocumented strict alignment constraints
 * [JDK-8307335](https://bugs.openjdk.org/browse/JDK-8307335): MemoryLayout.structLayout uses undocumented strict alignment constraints (**CSR**)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [592a3db6](https://git.openjdk.org/jdk/pull/13770/files/592a3db63b21cece89421dc673796b9cff84c472)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13770/head:pull/13770` \
`$ git checkout pull/13770`

Update a local copy of the PR: \
`$ git checkout pull/13770` \
`$ git pull https://git.openjdk.org/jdk.git pull/13770/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13770`

View PR using the GUI difftool: \
`$ git pr show -t 13770`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13770.diff">https://git.openjdk.org/jdk/pull/13770.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13770#issuecomment-1532651691)

### CODE REVIEW

**Comment by minborg** [comment]
Isn't this true also for sequence layouts where, for example, the second element might be misaligned for `sequenceLayout(2, structLayout(JAVA_INT, JAVA_SHORT))`? Or, is this already covered?

**Comment by mcimadamore** [comment]
> Isn't this true also for sequence layouts where, for example, the second element might be misaligned for `sequenceLayout(2, structLayout(JAVA_INT, JAVA_SHORT))`? Or, is this already covered?

That case already has some javadoc text:

https://cr.openjdk.org/~pminborg/panama/21/v2/javadoc/api/java.base/java/lang/foreign/MemoryLayout.html#sequenceLayout(long,java.lang.foreign.MemoryLayout)

(both factories throw if the alignment of the element layout is bigger than its size). It seems like I only forgot to add extra doc for structs.

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 1]
```diff
@@ -739,6 +739,28 @@ static SequenceLayout sequenceLayout(MemoryLayout elementLayout) {
      * @return a struct layout with the given member layouts.
      * @throws IllegalArgumentException if the sum of the {@linkplain #bitSize() bit sizes} of the member layouts
      * overflows.
+     * @throws IllegalArgumentException if a member layouts in {@code elements} occurs at an offset (relative to the start
+     * of the struct layout) which is not compatible with its alignment constraint.
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:743]:
```suggestion
     * @throws IllegalArgumentException if a member layout in {@code elements} occurs at an offset (relative to the start
     * of the struct layout) which is not compatible with its alignment constraint.
```

**Comment by mcimadamore** [comment]
> > Isn't this true also for sequence layouts where, for example, the second element might be misaligned for `sequenceLayout(2, structLayout(JAVA_INT, JAVA_SHORT))`? Or, is this already covered?
> 
> That case already has some javadoc text:
> 
> https://cr.openjdk.org/~pminborg/panama/21/v2/javadoc/api/java.base/java/lang/foreign/MemoryLayout.html#sequenceLayout(long,java.lang.foreign.MemoryLayout)
> 
> (both factories throw if the alignment of the element layout is bigger than its size). It seems like I only forgot to add extra doc for structs.

Actually, now that I look at it, it seems that the exception we raise is not correct. I think we need to throw if the element size is not a multiple of the alignment. That is, the size determines the stride, so that has to be a multiple of the alignment, otherwise it won't work.

**Review by JornVernee — APPROVED** [review]
Nice api note!

**Comment by JornVernee** [comment]
> Actually, now that I look at it, it seems that the exception we raise is not correct. I think we need to throw if the element size is not a multiple of the alignment. That is, the size determines the stride, so that has to be a multiple of the alignment, otherwise it won't work.

Hmm, I guess this doesn't work out if we have an element layout which has a size that is not a multiple of its alignment constraint (i.e. it lacks trailing padding). For instance a struct of size 12, with alignment 8. The second element of a sequence with that struct layout would not be aligned.

I think we went back-and-forth on the need to enforce trailing padding in the `structLayout` factory a few times.

**Comment by mcimadamore** [comment]
> > Actually, now that I look at it, it seems that the exception we raise is not correct. I think we need to throw if the element size is not a multiple of the alignment. That is, the size determines the stride, so that has to be a multiple of the alignment, otherwise it won't work.
> 
> Hmm, I guess the current check/assertion doesn't work out if we have an element layout which has a size that is not a multiple of its alignment constraint (i.e. it lacks trailing padding). For instance a struct of size 12, with alignment 8. The second element of a sequence with that struct layout would not be aligned.
> 
> I think we went back-and-forth on the need to enforce trailing padding in the `structLayout` factory a few times.

I think the rationale is: we can't enforce, in general, that a layout size is a multiple of its alignment. Because, for instance, we might want to construct "hyper-aligned" layouts (e.g. a struct layout aligned at page boundary. That's fine.

But I think when you use a layout as a sequence element, then we should check that the thing makes sense - e.g. that the element layout has a size that is indeed a multiple of the alignment. Otherwise it just seems like a "bad" sequence.

**Comment by mcimadamore** [comment]
> Actually, now that I look at it, it seems that the exception we raise is not correct. I think we need to throw if the element size is not a multiple of the alignment. That is, the size determines the stride, so that has to be a multiple of the alignment, otherwise it won't work.

If that's ok, I'd like to address this other issue in a separate PR/CSR, for clarity.

**Pull request closed without merging**